tests/scripts/generate_test_code.py

```diff
@@ -0,0 +1,1137 @@
+
+
+def gen_dispatch(name, dependencies):
+    """
+    Test suite code template main_test.function defines a C function
+    array to contain test case functions. This function generates an
+    initializer entry for a function in that array. The entry is
+    composed of a compile time check for the test function
+    dependencies. At compile time the test function is assigned when
+    dependencies are met, else NULL is assigned.
+
+    :param name: Test function name
+    :param dependencies: List of dependencies
+    :return: Dispatch code.
+    """
+    if dependencies:
+        preprocessor_check = gen_dependencies_one_line(dependencies)
+        dispatch_code = '''
+{preprocessor_check}
+    {name}_wrapper,
+#else
+    NULL,
+#endif
+'''.format(preprocessor_check=preprocessor_check, name=name)
+    else:
+        dispatch_code = '''
+    {name}_wrapper,
+'''.format(name=name)
+
+    return dispatch_code
+
+
+def parse_until_pattern(funcs_f, end_regex):
+    """
+    Matches pattern end_regex to the lines read from the file object.
+    Returns the lines read until end pattern is matched.
+
+    :param funcs_f: file object for .function file
+    :param end_regex: Pattern to stop parsing
+    :return: Lines read before the end pattern
+    """
+    headers = '#line %d "%s"\n' % (funcs_f.line_no + 1, funcs_f.name)
+    for line in funcs_f:
+        if re.search(end_regex, line):
+            break
+        headers += line
+    else:
+        raise GeneratorInputError("file: %s - end pattern [%s] not found!" %
+                                  (funcs_f.name, end_regex))
+
+    return headers
+
+
+def validate_dependency(dependency):
+    """
+    Validates a C macro and raises GeneratorInputError on invalid input.
+    :param dependency: Input macro dependency
+    :return: input dependency stripped of leading & trailing white spaces.
+    """
+    dependency = dependency.strip()
+    if not re.match(C_IDENTIFIER_REGEX, dependency, re.I):
+        raise GeneratorInputError('Invalid dependency %s' % dependency)
+    return dependency
+
+
+def parse_dependencies(inp_str):
+    """
+    Parses dependencies out of inp_str, validates them and returns a
+    list of macros.
+
+    :param inp_str: Input string with macros delimited by ':'.
+    :return: list of dependencies
+    """
+    dependencies = [dep for dep in map(validate_dependency,
+                                       inp_str.split(':'))]
+    return dependencies
+
+
+def parse_suite_dependencies(funcs_f):
+    """
+    Parses test suite dependencies specified at the top of a
+    .function file, that starts with pattern BEGIN_DEPENDENCIES
+    and end with END_DEPENDENCIES. Dependencies are specified
+    after pattern 'depends_on:' and are delimited by ':'.
+
+    :param funcs_f: file object for .function file
+    :return: List of test suite dependencies.
+    """
+    dependencies = []
+    for line in funcs_f:
+        match = re.search(DEPENDENCY_REGEX, line.strip())
```

Comment:
This function is not meant to be validating individual macros. It only parses suite dependency data.  Hence, the regex `DEPENDENCY_REGEX` is named after terminology used in the format specification.
Called function `parse_dependencies` extracts each macro from the data and validates it. The [`C_IDENTIFIER_REGEX`](https://github.com/ARMmbed/mbedtls/pull/930/files/f416494590082ee23340ce325980145aabc9ff42#diff-5801853c8d79838f04882c360fe720b3R187) regex validates the macro in function [`validate_dependency`](https://github.com/ARMmbed/mbedtls/pull/930/files/f416494590082ee23340ce325980145aabc9ff42#diff-5801853c8d79838f04882c360fe720b3R379)

Reply:
Alright, then it seems the checks were not sufficient because the following patch makes compilation fail as I described above:
```
diff --git a/tests/suites/test_suite_base64.function b/tests/suites/test_suite_base64.function
index 3a8bf43..c796617 100644
--- a/tests/suites/test_suite_base64.function
+++ b/tests/suites/test_suite_base64.function
@@ -3,8 +3,7 @@
 /* END_HEADER */

 /* BEGIN_DEPENDENCIES
- * depends_on:MBEDTLS_BASE64_C
- * END_DEPENDENCIES
+ * depends_on:MBEDTLS_BASE64_C END_DEPENDENCIES
  */

 /* BEGIN_CASE */
```

Reply:
Thanks for identifying this. I don't understand why `re.match` should behave like `re.search` in this case. A `$` at the end fixes this issue.

Reply:
`re.match(regex, string)` matches if `string` starts with a match for `regex`. `re.search` matches if `string` contains a match for `regex`. There is no function that matches if `string` is exactly a match for `regex`, use `\Z` in the regex for that.